torch/csrc/autograd/VariableTypeUtils.h

```diff
@@ -148,12 +148,12 @@ inline std::vector<SavedVariable> make_saved_variable_list(TensorList tensors) {
 }
 
 inline Tensor as_variable(Tensor tensor) {
-  return make_variable(std::move(tensor), /*requires_grad=*/false);
+  return tensor.is_variable() ? tensor : make_variable(std::move(tensor), /*requires_grad=*/false);
 }
 
 inline std::vector<Tensor> as_variable(TensorList tl) {
   return fmap(tl, [](const Tensor& t) -> Tensor {
-      return make_variable(t, /*requires_grad=*/false);
+      return t.is_variable() ? t : make_variable(t, /*requires_grad=*/false);
```

Comment:
We changed the implementation of `as_variable()` here, because there is no guarantee that the tensors returned from out-of-place ATen ops are not Variables. For example, `coalesce()` (called from `VariableType::coalesce()`), `_embedding_bag_cpu()` (called from `VariableType::_embedding_bag()`) can return its input or tensors created using input's options, and the inputs to those functions can potentially be Variables, since we don't do unwrap on the inputs anymore when we pass them into the ATen ops in VariableType.cpp (see https://github.com/pytorch/pytorch/pull/17072/files#diff-57fa25b47cf60508d3eb2719bc7ccb4aR183).

Reply:
unless I'm misunderstanding something, calling `tensor_data()` in unpack is just an optimization, no?  Like, you can make the same argument about saving variables with autograd meta or not.  And I'd expect optimization decision to be driven by benchmarks.

Reply:
I ran a simple benchmark (small `nn.LSTM` model) on the two versions (calling vs. not calling `tensor_data()` in `unpack()`), and here are the results:

Benchmark code:
```python
import timeit

print(timeit.timeit('''
output, hn = rnn(input, (h0, c0))''',
setup='''
import torch;
rnn = torch.nn.LSTM(input_size=10, hidden_size=20, num_layers=2);
input = torch.randn(5, 3, 10);
h0 = torch.randn(2, 3, 20);
c0 = torch.randn(2, 3, 20);
# warm up
output, hn = rnn(input, (h0, c0));
output, hn = rnn(input, (h0, c0));
output, hn = rnn(input, (h0, c0))''', number=1000))
```

Results:
```
master: 1.1164182517677546
not calling tensor_data(): 1.1378995887935162
calling tensor_data(): 1.224072452634573
```
The "calling tensor_data()" version consistently causes ~10% overhead compared to master, and I think we should have the optimization of not calling `tensor_data()` in `unpack()`.

Reply:
These changes are only temporary - after we make all tensors variables, these `as_variable()` functions are not needed anymore and will be removed.